shop.cpp

```diff
@@ -3,24 +3,57 @@
 
 Shop::Shop(int mMaxChairs)
     :mMaxChairs(mMaxChairs)
-    ,mTheBarber(this)
-    ,mBarberSemaphore(0)
-    ,mCustomersSemaphore(0)
 {
+    mBarberSemaphore = new Semaphore(1);
+    mCustomersSemaphore = new Semaphore(0);
+    mTheBarber = new Barber(this, mBarberSemaphore, mCustomersSemaphore);
+}
+
+Shop::~Shop()
+{
+    if (mTheBarber != nullptr)
+    {
+        delete mTheBarber;
```

Comment:
How can you be certain that the barber thread has exited before you delete the barber object?